### ISSUE TITLE
[BUG] 'Output' object has no attribute 'runs_before'

### ISSUE THREAD

**Issue by ajsalow** [flyteorg/flyte#957]
**Describe the bug**
I've been unable to create a simple test that reproduces the error so I'll show an example from my code that is producing it.

Note: the `_top_worker` functions are dynamic tasks that call subtasks.
```python
cache_node = create_node(cache_parquet_top_worker,country=country)

if transforms is None:
    transforms = []

embed_node = create_node(embed_top_worker, country=country, transforms=transforms)
print(type(cache_node))
cache_node.runs_before(embed_node)
```

This code is executed twice when a workflow runs. The first time the type of cache node is `<class 'flytekit.core.node.Node'>` the second time it is. `<class 'flytekit.core.promise.create_task_output.<locals>.Output'>`

The `Output` object has not attribute `runs_before`. The `Output` class in question is found [here](https://github.com/flyteorg/flytekit/blob/e5302c428eaeb4d0c3220a6e6ee41cd554bfb77f/flytekit/core/promise.py#L454).

A comment in the code mentions asks the question "Should this class be part of the Interface?" The `interface.Interface.Output` class has a runs_before method that does nothing.

```python
def runs_before(self, *args, **kwargs):
  """
  This is a placeholder and should do nothing. It is only here to enable local execution of workflows
  where runs_before is manually called.
  """
```

When I was trying to recreate this in a unit test I noticed the type of the node was always <class 'flytekit.core.interface.Interface.__init__.<locals>.Output'>, which is why it was not failing. 

My question is, am I creating an invalid workflow? If not, then should the `promise.output` also have a do nothing `runs_before` method or should it use the same output class in interface.py?

**Expected behavior**
Attribute error not thrown.

**Comment by ajsalow** [comment]
Update: I was able to write a unit test that reproduces.

```python
def test_runs_before():
    @task
    def t2(a: str, b: str) -> str:
        return b + a

    @task()
    def sleep_task(a: int) -> str:
        a = a + 2
        return "world-" + str(a)

    @dynamic
    def my_subwf(a: int) -> (typing.List[str], int):
        s = []
        for i in range(a):
            s.append(sleep_task(a=i))
        return s, 5

    @workflow
    def my_wf(a: int, b: str) -> (str, typing.List[str], int):
        t2_node = create_node(t2,a=b, b=b)
        subwf_node = create_node(my_subwf, a=a)
        print(type(subwf_node))
        subwf_node.runs_before(t2_node)
        return t2_node.o0, subwf_node.o0, subwf_node.o1

    my_wf(a=5, b="hello")
```

**Comment by kumare3** [comment]
@wild-endeavor fyi!

**Comment by wild-endeavor** [comment]
Thoughts? https://github.com/flyteorg/flytekit/pull/465/files

**Comment by ajsalow** [comment]
I think we also need to override `__rshift__`

**Comment by wild-endeavor** [comment]
@ajsalow this was released under v0.19.0b1 - mind giving it a try?  If it works we can close out this ticket.

**Comment by ajsalow** [comment]
Tested and confirmed that the above issue is fixed. However, `runs_before` still doesn't work as expected when run locally. 

For example the following code should have `mult_node` run first and the final value of var should be 12, but it is 14.
```python
var = 5


@task
def add_2() -> int:
    print(f"add time:{datetime.now()}")

    global var
    var+=2
    return var

@task()
def mult_2() -> int:
    print(f"mult time:{datetime.now()}")

    global var
    var*=2
    return var

@workflow
def my_wf() -> (int, int):
    add_node = create_node(add_2)
    mult_node = create_node(mult_2)
    mult_node.runs_before(add_node)
    return add_node.o0, mult_node.o0

add, mult = my_wf()
```
Output
```bash
2021-05-14 10:32:15,334-flytekit-INFO$ Invoking __main__.add_2 with inputs: {}
add time:2021-05-14 10:32:15.334374
2021-05-14 10:32:15,334-flytekit-INFO$ Task executed successfully in user level, outputs: 7
2021-05-14 10:32:15,334-flytekit-INFO$ Invoking __main__.mult_2 with inputs: {}
mult time:2021-05-14 10:32:15.335069
2021-05-14 10:32:15,335-flytekit-INFO$ Task executed successfully in user level, outputs: 14
```

`add_node` clearly ran before `mult_node`. This specific example doesn't translate to running in flyte because of the global variable. (Actually I did try running this on flyte just to see what would happen and it actually executes but gives grabage results. Maybe there should be check to enforce no globals in tasks or workflows) When running other examples in flyte `runs_before` does order node execution correctly. I think this can be fixed under a new issue or a small temporary fix would be to have `run_node`, when executed locally, log a warning that it will not reorder execution when running locally.
